File: 1422/maximum-score-after-splitting-a-string.py
class Solution:
    def maxScore(self, s: str) -> int:
        N = len(s) # 長度
        zero = 0 # 等一下也要
        one = 0 # 想找出全部的1
        for i in range(N):
            if s[i]=='1': one += 1 # 如果是1,先全部加起來
        # 現在就知道總共有幾個 one
        #print('一開始的時候,都在右邊,統計結果','zero', zero, 'one', one)
        ans = 0
        for i in range(N-1): # 要逐格去做修正, 「吐出」一格,看它是多少,就修正
            if s[i]=='0': # 吐出'0' 給左邊
                zero += 1 # 左邊多1個0
            if s[i]=='1': # 吐出'1' 給左邊, 右邊就少了1個'1'
                one -= 1 # 右邊少1個1
            #print('中間過程中,','zero', zero, 'one', one)
            ans = max( ans, zero+one) # 最好的答案、最大的值
        return ans

File: 1422/maximum-score-after-splitting-a-string.cpp
// LeetCode 1422. Maximum Score After Splitting a String
// 字串 s 裡有一堆 0 和 1，把它分「左右」兩半，左邊0的個數 + 右邊1的個數，合起來要最多。
class Solution {
public:
    int maxScore(string s) {
        int ans = 0, left0 = 0, right1 = 0; // 左邊0 和 右邊1
        for(int i=0; i<s.length(); i++) { // 先全部算到右邊
            if(s[i]=='1') right1++; // 先統計「右邊1」有幾個
        }
        for(int i=0; i<s.length()-1; i++) { // 再來，逐步增加「左邊」
            if(s[i]=='0') left0++; // 左邊多個1
            else if(s[i]=='1') right1--; // 右邊少個0
            ans = max(ans, left0 + right1);
        }
        return ans;
    }
};
